### DISCUSSION TITLE
How to log the filename of the file printed by cups?

### DISCUSSION THREAD

**Discussion by utt54321** [OpenPrinting/cups#418] [Q&A] [answered]
How to log the filename of the file printed by cups?
The filename can be like "/path/to/text.pdf".
I want to print this filename in /var/log/messages.
But the **struct cupsd_job_t** (schedule/job.h) is very complex for me to find a good point to do so.
Can any ace help me to figure out this stupid question?

**Comment by utt54321** [comment]
Please someone helps me.
q_q

**Comment by zdohnal** [comment]
Hi @utt54321 ,

afaik we don't save anywhere the file with absolute path - we only save the filename itself, under ```name``` in the ```cupsd_job_t``` - it is then used as a job title.

**Reply by utt54321** [reply]
Hi zdohnal,
1.But I put a txt file (like a.txt) to firefox, and than I use firefox to print it. I found the title is "firefox", where it is not the file's name. Why？
2.If there is nowhere saving absolute path of the printed file, how could cups find it and use it as inputfile? Is there any possibility to find the file's absolute path.

Very thankful for your helping.

**Reply by michaelrsweet** [reply]
@utt54321 The original filename is *not* part of the job ticket sent to cupsd. By default, if you "lp /path/to/filename", the default title (job-name) value is the base filename ("filename") but that can be overridden. Many applications will use the document title ("My Great Web Page", "2022 Q2 Finance Report", etc.) and if no title is supplied you'll see "Untitled".

**Comment by jsmeix** [comment]
Regarding what applications specify as job title
when submitting a print job to CUPS:

This is something that belongs to the individual application.

When the application is a browser:
I experienced a longer time ago that a browser used the URL
as job title which is not a good idea on systems that are used
by more than one user because a URL is personal information
(what sites you visit is personal information)
and the job title is not a secret value so users
may see job titles of other user's print jobs.

**Reply by utt54321** [reply]
Thank you, sir.
I understand what you say.
If I use a pdf editor to print, how could I know its absolute path in cups?

**Comment by jsmeix** [comment, marked as answer]
Regarding
```
path of the printed file, how could cups find it and use it as inputfile
```

When you print directly on command line via
```
# lp -d queue_name /path/to/file
```
the `lp` program knows /path/to/file
and submits the content of /path/to/file
as print job data via IPP protocol to the cupsd
which stores the print job data in its spool directory
as a new file named /var/spool/cups/d.....-...

So the original file name is not preserved
except e.g. explicitly provided as print job title
when you print via
```
# lp -d queue_name -t '/path/to/file' /path/to/file
```

**Reply by utt54321** [reply]
Sir，I am greatly thankful for your describing this process.
I have another question: I am looking for a way to log this  " /path/to/file " when it is submitted as print job data, where can I find it (where is the code?)

**Reply by jsmeix** [reply]
When you print via
```
# lp -d queue_name -t '/path/to/file' /path/to/file
```
the string "/path/to/file" from `-t '/path/to/file'`
is not submitted as print job data but as a print job option
which is here the job title option with value "/path/to/file".

What is submitted as print job data is the content of /path/to/file
(print job data is what basically "appears on the printed paper")
but the filename is not submitted so
```
# lp -d queue_name -t '/path/to/that' /path/to/this
```
submits a print job with the title option set to "/path/to/that"
while the content of /path/to/this is submitted as print job data.

When the cupsd runs with `LogLevel debug` set in /etc/cups/cupsd.conf
it logs the print job title as `argv[2]` in its log file
which is in my case (om openSUSE Leap 15.3) /var/log/cups/error_log
as a debug log line `D ...` that looks for example like
```
D [21/Jun/2022:12:34:56 +0200] [Job 123] argv[2]="/path/to/that"
```

Do you intend to change the cupsd code to let it log things differently?

**Reply by utt54321** [reply]
Oh, I start to understand.
1. If we use lp command with -t, we can set the title.
2. lp will not cpoy the absolute path of the file to the print job data, if without -t, it will set filename as the title.
3. The process of the whole printing is:
    user choose an application(like lp) to print /path/to/file ->the content of /path/to/file has been  treated as part of data in this print job -> the print job data with no any info of the absolute path of file, but the title can be name of the printed file if the title is not set -> transform the print job data to the cupsd via ppl -> cupsd then do the print job ......
4. So it is impossible to get the filename at anytime(because the title can be changed) or the absolute path of the file in the cups code.

Do I understand right?

Then I have another question:
the below code has argv[6] as filename, dose it have any relationship to the /path/to/file ?(I am very bad at looking and understanding code, so I am very confused about this argv[6]) Where these argvs come from and what they aim for?
https://github.com/OpenPrinting/cups/blob/7b4e1c8775c2793faf2d59dc306f84dcdf80b9fd/scheduler/job.c#L866-L922

**Reply by michaelrsweet** [reply]
@utt54321 No, the filename passed to the print filters is the spool filename, typically of the form "/var/spool/cups/dNNNNNN-nnn" where NNNNNN is the job ID number and nnn is the file/document number in the job (starting at 001).

**Reply by utt54321** [reply]
Hi michaelrsweet,
Are the four points I mentioned above right? (The really impotant thing is I want to confirm that in the whole process only the tiltle has chance to be the filename of printed file and there is nowhere storing the filename. So we can not log the filename and only can log the title.)

I am very very appreciate for your help.

**Reply by zdohnal** [reply]
@utt5432

AFAIK:

1) yes.
2) if ```-t``` is defined, the string which comes after it becomes the job title. If you don't define ```-t```, ```lp``` takes the filename without path as job title - ```lp -d test /etc/fstab``` -> job title is ```fstab``` - but as @jsmeix wrote, this depends on implementation of application which sends the file to ```cupsd```
3) yes in general - app basically transforms job options (such is a job title) into IPP attributes, adds the contents of file to be printed and sends it all as IPP request(s) (PrintJob or CreateJob+SendDocument) to cupsd. cupsd saves it in /var/spool/cups during job creation and then starts to process it - so the absolute path of the file you want to print is not needed in cupsd. Actually the applications like evince don't even send the same file to cupsd at all - they generate a copy of the document on their own via generators like cairo, so strictly speaking, the file which cupsd gets can be a different file than the one which is under the absolute path.
4) Yes. You can only get the job title, which format depends on how an application which sends the file to cupsd defines it. ```lp``` uses the filename f.e. ```fstab``` but that's implementation specific.

**Reply by utt54321** [reply]
Very very thankful for your help.

**Reply by utt54321** [reply]
I have an idea that I can log filename in the cupsDoFileRequest function:
https://github.com/OpenPrinting/cups/blob/8456ca2bdde7922310d551633a91a95fd6674230/cups/request.c#L41-L78
I changed the DEBUG_printf to be printf:
`printf(("cupsDoFileRequest(http=%p, request=%p(%s), resource=\"%s\", filename=\"%s\")", (void *)http, (void *)request, request ? ippOpString(request->request.op.operation_id) : "?", resource, filename));`
But I use lp print a file, there is no log in the /var/log showing "cupsDoFileRequest(http=......", why?

**Reply by michaelrsweet** [reply]
@utt54321 The DEBUG_xxx macros only enable debug logging for the client side. They are not normally compiled in (you need to configure with the "--enable-debug" option) and add a lot of overhead to the libcups APIs. And even if you do build CUPS with that configure option, you need to enable logging with the `CUPS_DEBUG_LOG` environment variable.

Basically, what you want is not possible.

**Reply by utt54321** [reply]
Hi michaelrsweet，
I have found a lot of applications supports printing. Firefox and featherpad support printing, but firefox's deb source dose not rely on any cups libs and featherpad seems to use QT api to support printing. They seem to work with cups doing printing job, but I cannot understand that how they send ipp request without using cups libs. How could they work so well?

**Reply by zdohnal** [reply]
QT implements the communication with cupsd via CUPS API, so it is QT who depends on libcups - featherpad depends on libcups transitively, via QT. 
Firefox uses runtime linking - it defines dummy methods with names matching to libcups API and once Firefox is started, it checks for libcups, and if it exists, loads the library and its functions replace the dummy ones.
This way Firefox doesn't have to depend on libcups - several more applications do that with other libs (f.e. Vim doesn't require Python3 libs to narrow its deps...)
